### ISSUE TITLE
setenv()/unsetenv() calls break libmisc/env.c "newenvc" semantics, causes segmentation fault

### ISSUE THREAD

**Issue by rpjday** [shadow-maint/shadow#233]
In shadow 4.2.1 (and visually everything since then), libmisc/env.c assumes that newenvp and newenvc stay in sync, such that everything up to newenvp[newenvc-1] is valid, and newenvp[newenvc] is NULL.

Apparently, calls made by applications that use shadow library that get ahold of this array can break the semantics of newenvc, such that, at some point, when a call to setenv() or unsetenv() is made, after which the routine addenv() is called, newenvp[newenvc-1] is now NULL, something addenv() is not prepared for and which causes a segmentation fault.

In short, env.c is unprepared for other code to manipulate the environment array while not updating the internal "newenvc" counter -- this is, in fact, exactly what is happening in src/login.c (among other places).

**Comment by hallyn** [comment]
I'm afraid I'm not seeing the src/login.c sequence that manipulates newenvp.  Can you show how you triggered the segfault?
